blockly-ws-search/src/WorkspaceSearch.js

```diff
@@ -0,0 +1,660 @@
+    const metrics = this.workspace_.getMetrics();
+    if (this.workspace_.RTL) {
+      this.htmlDiv_.style.left = metrics.absoluteLeft + 'px';
+    } else {
+      if (metrics.toolboxPosition == Blockly.TOOLBOX_AT_RIGHT) {
+        this.htmlDiv_.style.right = metrics.toolboxWidth + 'px';
+      } else {
+        this.htmlDiv_.style.right = '0';
+      }
+    }
+    this.htmlDiv_.style.top = metrics.absoluteTop + 'px';
+  }
+
+  /**
+   * Handles input value change in search bar.
+   * @private
+   */
+  onInput_() {
+    if (this.searchOnInput) {
+      const inputValue = this.inputElement_.value;
+      if (inputValue !== this.searchText_) {
+        this.setSearchText_(inputValue);
+        this.searchAndHighlight(this.preserveSelected);
+      }
+    }
+  }
+
+  /**
+   * Handles a key down for the search bar.
+   * @param {KeyboardEvent} e The key down event.
+   * @private
+   */
+  onKeyDown_(e) {
+    if (e.key === 'Escape') {
+      this.close();
+    } else if (e.key === 'Enter') {
+      if (this.searchOnInput) {
+        this.next();
+      } else {
+        this.setSearchText_(this.inputElement_.value);
+        this.searchAndHighlight(this.preserveSelected);
+      }
+    }
+  }
+
+  /**
+   * Opens the search bar when Control F or Command F are used on the workspace.
+   * @param {KeyboardEvent} e The key down event.
+   * @private
+   */
+  onWorkspaceKeyDown_(e) {
+    // TODO: Look into handling keyboard shortcuts on workspace in Blockly.
+    if ((e.ctrlKey || e.metaKey) && e.key === "f") {
+      this.open();
+      e.preventDefault();
+    }
+  }
+
+  /**
+   * Selects the previous block.
+   */
+  previous() {
+    if (!this.blocks_.length) {
+      return;
+    }
+    this.setCurrentBlock_(this.currentBlockIndex_ - 1);
+  }
+
+  /**
+   * Selects the next block.
+   */
+  next() {
+    if (!this.blocks_.length) {
+      return;
+    }
+    this.setCurrentBlock_(this.currentBlockIndex_ + 1);
+  }
+
+  /**
+   * Sets the placeholder text for the search bar text input.
+   * @param {string} placeholderText The placeholder text.
+   */
+  setSearchPlaceholder(placeholderText) {
+    this.textInputPlaceholder_ = placeholderText;
+    if (this.inputElement_) {
+      this.inputElement_.setAttribute('placeholder', this.textInputPlaceholder_);
+    }
+  }
+
+  /**
+   * Sets search text.
+   * @param {string} text
+   * @protected
+   */
+  setSearchText_(text) {
+    this.searchText_ = text.trim();
+  }
+
+  /**
+   * Changes the currently "selected" block and adds extra highlight.
+   * @param {number} index Index of block to set as current. Number is wrapped.
+   * @protected
+   */
+  setCurrentBlock_(index) {
+    if (!this.blocks_.length) {
+      return;
+    }
+    if (this.currentBlock_) {
+      this.unhighlightCurrentSelection_(this.currentBlock_);
+    }
+    this.currentBlockIndex_ =
+        (index % this.blocks_.length + this.blocks_.length) %
+        this.blocks_.length;
```

Comment:
Simply: ``index % this.blocks_.length``

Reply:
That simplification doesn't work when index is negative (which we are trying to handle).

Reply:
Sorry, didn't see this before I changed. Lmk and I will update.

Reply:
Changed back